src/cs/Ssh.Tcp/Services/PortForwardingService.cs

```diff
@@ -557,33 +594,40 @@ public PortForwardingService(SshSession session)
 	}
 
 	private async Task<int?> StartForwardingAsync(
-		IPEndPoint localEndPoint,
+		IPAddress localIPAddress,
+		int remotePort,
 		CancellationToken cancellation)
 	{
 		cancellation.ThrowIfCancellationRequested();
 
 		if (AcceptLocalConnectionsForForwardedPorts)
 		{
-			// The initially requested local endpoint uses the remote port, but it may change
+			// The local port is initially set to the remote port, but it may change
 			// when starting forwarding, if there was a conflict.
-			var remotePort = localEndPoint.Port == 0 ? null : (int?)localEndPoint.Port;
+			int localPort = remotePort;
 
 			var forwarder = new LocalPortForwarder(
 				this,
 				Session,
 				PortForwardChannelType,
-				localEndPoint.Address,
-				localEndPoint.Port);
+				localIPAddress,
+				localPort,
+				remoteHost: null,
+				remotePort == 0 ? null : remotePort);
 
 			await forwarder.StartForwardingAsync(cancellation).ConfigureAwait(false);
+			localPort = forwarder.LocalPort;
 
-			var forwardingEndPoint = new IPEndPoint(localEndPoint.Address, forwarder.LocalPort);
+			// The remote port is the port received in the message from the other side,
+			// so the forwarder is indexed on that port number, rather than the local port,
+			// unless the request was to choose a random port.
```

Comment:
It seems we may be mixing two namespaces here: remote ports and local ports, and as local ports may be different from remote ports, there may be conflicts. For example
1. Request to choose a random port. Local port 22345 was chosen and used as the key
2. Request to forward with remote port 22345. Some free local port was chosen but we still try to use port 22345 because it's the remote port - Conflict.

Reply:
That second request should be rejected because that port number is already forwarded. It doesn't matter whether on the first request it was chosen randomly or specifically.

But you're right that there is no check for that when initially handling the request. I'll add a test case and fix the issue.

Reply:
Added a check and test case.